fv3gfs/util/halo_data_transformer.py

```diff
@@ -0,0 +1,914 @@
+    dtype: Any
+
+
+# ------------------------------------------------------------------------
+# Simple pool of streams to lower the driver pressure
+# Use _pop/_push_stream to manipulate the pool
+
+STREAM_POOL: List["cp.cuda.Stream"] = []
+
+
+def _pop_stream() -> "cp.cuda.Stream":
+    if len(STREAM_POOL) == 0:
+        return cp.cuda.Stream(non_blocking=True)
+    return STREAM_POOL.pop()
+
+
+def _push_stream(stream: "cp.cuda.Stream"):
+    STREAM_POOL.append(stream)
+
+
+# ------------------------------------------------------------------------
+# Indices array
+
+# Keyed cached - key is a str at the moment to go around the fact that
+# a slice is not hashable. getting a string from
+# Tuple(slices, rotation, shape, strides, itemsize) e.g. # noqa
+# str(Tuple[Any, int, Tuple[int], Tuple[int], int]) # noqa
+INDICES_CACHE: Dict[str, "cp.ndarray"] = {}
+
+
+def _build_flatten_indices(
+    key,
+    shape,
+    slices: Tuple[slice],
+    dims,
+    strides,
+    itemsize: int,
+    rotate: bool,
+    rotation: int,
+) -> "cp.ndarray":
+    """Build an array of indexing from a slice & memory description to build an indexation into the "flatten" memory.
+    
+    Go from a memory layout (strides, itemsize, shape) and slices into it to a
+    single array of indices. We leverage numpy iterator and calculate from the multi_index
+    using memory layout the index into the original memory buffer.
+    """
+
+    # Have to go down to numpy to leverage indices calculation
+    arr_indices = np.empty(shape, dtype=np.int32, order="C")[slices]
+
+    # Get offset from first index
+    offset_dims = []
+    for s in slices:
+        offset_dims.append(s.start)
+    offset_to_slice = sum(np.array(offset_dims) * strides) // itemsize
+
+    # Flatten the index into an indices array
+    with np.nditer(
+        arr_indices, flags=["multi_index"], op_flags=["writeonly"], order="K",
+    ) as it:
+        for array_value in it:
+            offset = sum(np.array(it.multi_index) * strides) // itemsize
+            array_value[...] = offset_to_slice + offset
+
+    if rotate:
+        # sending data across the boundary will rotate the data
+        # n_clockwise_rotations times, due to the difference in axis orientation.
+        # Thus we rotate that number of times counterclockwise before sending,
+        # to get the right final orientation. We apply those rotations to the
+        # indices here to prepare for a straightforward copy in cu kernel
+        arr_indices = rotate_scalar_data(arr_indices, dims, cp, -rotation)
+    return cp.asarray(arr_indices.flatten(order="C"))
+
+
+# ------------------------------------------------------------------------
+# HaloDataTransformer helpers
+
+
+def _slices_size(slices: Tuple[slice, ...]) -> int:
+    """Compute linear size from slices."""
+    length = 1
+    for s in slices:
+        assert s.step is None
+        length *= abs(s.start - s.stop)
+    return length
+
+
+@dataclass
+class HaloExchangeData:
```

Comment:
Suggest renaming this to `HaloExchangeSpec` because it does not contain data, only a specification about data. Not sure how to make it clearer though that this is about a specific pair of ranks and not about a full halo exchange.